src/repo.rs

```diff
@@ -59,6 +59,42 @@ impl Repository {
         }
     }
 
+    /// Find and open an existing repository, with additional options.
+    ///
+    /// If flags contains REPOSITORY_OPEN_NO_SEARCH, the path must point
+    /// directly to a repository; otherwise, this may point to a subdirectory
+    /// of a repository, and `open_ext` will search up through parent
+    /// directories.
+    ///
+    /// If flags contains REPOSITORY_OPEN_CROSS_FS, the search through parent
+    /// directories will not cross a filesystem boundary (detected when the
+    /// stat st_dev field changes).
+    ///
+    /// If flags contains REPOSITORY_OPEN_BARE, force opening the repository as
+    /// bare even if it isn't, ignoring any working directory, and defer
+    /// loading the repository configuration for performance.
+    ///
+    /// ceiling_dirs specifies a list of paths (separated by the platform path
+    /// list separator) that the search through parent directories will stop
+    /// before entering.  Use the functions in std::env to construct or
+    /// manipulate such a path list.
+    pub fn open_ext<P: AsRef<Path>>(path: P,
+                                    flags: RepositoryOpenFlags,
+                                    ceiling_dirs: Option<&str>)
```

Comment:
> It's not necessarily that hard to do something like env::split_path(&val), though, right? That returned iterator would satisfy the requirement if we have the right generic bounds.

And then `open_ext` would immediately call `env::join_paths` on it; that seems like a waste. :)

Since we have functions in `std::env` to do the conversion in either direction, and since libgit2 wants the joined form, it seems like we should either accept the joined form and have callers call `env::join_paths` if they have the split form, or accept both and call `env::join_paths` ourselves if given the split form.  Personally, I think the former seems simplest: accept an IntoCString and have the caller use `env::join_paths` if needed.